app/qml/form/editors/MMFormPhotoEditor.qml

```diff
@@ -160,14 +161,43 @@ MMFormPhotoViewer {
   }
 
   function callbackOnFormSaved() {
-    if ( internal.imageSourceToDelete ) {
-      __inputUtils.removeFile( internal.imageSourceToDelete )
-      internal.imageSourceToDelete = ""
+    clearDeletedPhotos()
+  }
+
+  function callbackOnFormCanceled( formState ) {
+    // check different form states to handle existing and deleted photos when cancelling
+    if ( formState === "add" ) {
+      // in "add" mode, if there's a value and we're cancelling, add it to deletion queue
+      if ( root._fieldValue ) {
+        let existingPhotoPath = __inputUtils.getAbsolutePath( root._fieldValue, internal.prefixToRelativePath )
+        internal.imageSourcesToDelete.push( existingPhotoPath )
+      }
+
+      clearDeletedPhotos()
+    }
+    else if ( formState === "edit" ) {
+      // in "edit" mode, we need special handling:
+      // 1. if there was an original value, remove it from deletion queue
+      if ( root._fieldOriginalValue ) {
+        var originalAbsPath = __inputUtils.getAbsolutePath( root._fieldOriginalValue, internal.prefixToRelativePath )
+        var idx = internal.imageSourcesToDelete.indexOf( originalAbsPath )
+        if ( idx !== -1 ) {
+          internal.imageSourcesToDelete.splice( idx, 1 )
+        }
+      }
+      // 2. add current value to deletion queue if it exists and it's not the original value
+      if ( root._fieldValue && root._fieldValue !== root._fieldOriginalValue ) {
+        let existingPhotoPath = __inputUtils.getAbsolutePath( root._fieldValue, internal.prefixToRelativePath )
+        internal.imageSourcesToDelete.push( existingPhotoPath )
+      }
```

Comment:
I still have mixed feelings about this code. I would feel much safer if it was in c++ and covered with unit tests.. I am afraid this can explode in ways we do not imagine now. This is just too much javascript/qml logic for something this critical. My suggestion would be this:
 1. we merge just the renaming fix (`callbackOnFormCanceled`)
 2. we close the other work and keep that in mind (reference) for the photo refactoring - there we will put all of this code into cpp and cover with unit tests..